gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java

```diff
@@ -278,6 +284,39 @@ public Response revoke(String token) {
     return resp;
   }
 
+  @GET
+  @Path(PUBLIC_KEY_PATH)
+  @Produces ({APPLICATION_JSON})
+  public Response publicKey() {
+    GatewayServices services = (GatewayServices) request.getServletContext()
+        .getAttribute(GatewayServices.GATEWAY_SERVICES_ATTRIBUTE);
+
+    KeystoreService ks = services.getService(ServiceType.KEYSTORE_SERVICE);
+
+    try {
+      PublicKey publicKey = ks.getSigningKeystore().getCertificate(getSigningKeyAlias()).getPublicKey();
+      publicKey.getAlgorithm();
+      byte[] pubKeyBytes = publicKey.getEncoded();
+      String base64Key = Base64.encodeBase64String(pubKeyBytes);
+      Map<String, String> map = new HashMap<>();
+      map.put(PUBLIC_KEY, base64Key);
+      map.put(ALGORITHM, publicKey.getAlgorithm());
+
+      String jsonResponse = JsonUtils.renderAsJsonString(map);
+      return Response.ok().entity(jsonResponse).build();
+    } catch (KeyStoreException | KeystoreServiceException e) {
+      return Response.serverError().build();
```

Comment:
We should at least log the exception message here in the gateway log.
I'm always conflicted about how much detail we should return to the client in such cases.

Reply:
Will add a log line.
Yup - what level of information goes back to the client is always problematic. In this particular case, I doubt the client learns much from the error going back (There's no parameters etc for the client to get wrong which they could try fixing), and this requires a server side fix.

Reply:
I agree with that assessment of the response content.